### DISCUSSION TITLE
Support assertions on arguments in the wrapper

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#2077] [Ideas]
Assertions fail brutally when translated by Pyccel. It would be nice if assertions that are used to validate arguments could be collected and managed in the wrapper. This would allow the code to fail cleanly if the input is not validate.

**Comment by yguclu** [comment]
Most assertions should translate naturally to C, given that `assert.h` is part of the C standard library.

As for Fortran, most compilers should allow us to call the corresponding C functions through a minimal wrapper. For example, [here](https://github.com/selalib/selalib/tree/main/src/low_level_utilities/assert) is how the SeLaLib library handles asserts.

**Reply by EmilyBourne** [reply]
> Most assertions should translate naturally 

I agree, this is what we do currently. But the difference between handling them in the wrapper and handling them via `assert.h` is that wrapper assertions can be written to not crash the Python instance.

For example suppose I wrote a class that is not very general (this is something I have done in Fortran before in order to create a buffer whose size is known at compile time for a GPU implementation):
```python
class BSplines:
    def __init__(self, knots : 'float[:]', degree : int, periodic : bool, uniform : bool):
        assert degree < 5
        ...
```
and then handed it off to a user who was not aware of this restriction. If they try to use a large degree then their code will crash:
![image](https://github.com/user-attachments/assets/bc18f4d7-d811-4f58-931a-01585dc3bd17)

But by simply adding 4 lines to the wrapper:
```c
    if (!(degree < 7)) {
        PyErr_SetString(PyExc_AssertionError, "degree < 7");
        return -1;
    }
```
we can get a more Pythonic behaviour:
![image](https://github.com/user-attachments/assets/4bb99ea0-4169-4802-b420-069da9aae248)

If someone is using their own code then I don't think the current behaviour is problematic, however if Pyccel is used to accelerate a library (e.g. Struphy) then it becomes more important to ensure that the generated library is user friendly. I suspect this will become more important as we support more code, as users of libraries using Pyccel are more likely to interact with the Pyccelised code directly.

**Reply by yguclu** [reply]
Sorry, I was confused by the first sentence in the PR description, "Assertions fail brutally when translated by Pyccel". To me that meant that Pyccel could not **translate** assertions.

Are you suggesting to move the assert checking to the wrapper completely? Or simply to handle failures there and then raise a nice Python exception?

**Reply by EmilyBourne** [reply]
My suggestion was to keep the low level translation as is, but additionally add statements such as the 4 lines listed above for the assertions concerning arguments.

I don't know if we can catch all errors in the wrapper? This could be especially difficult if they are coming from Fortran.

Handling assertions explicitly allows us to print the condition causing the failure.

**Reply by EmilyBourne** [reply]
I think we should leave the low level language as is in case the translation is used without the wrapper.

I think assertions are deactivated in release mode in C so handling this in the wrapper could be safer

**Reply by yguclu** [reply]
I totally agree with your last message. The one thing I am not so sure about is distinguishing between the assertions that concern the arguments and those that do not. Can you do that in a general way?

**Reply by EmilyBourne** [reply]
Yes this should be pretty easy:
```python
assertions_for_wrapper = [a for a in starting_assertions if all(v.is_argument for v in a.get_attribute_nodes(Variable))]
```

**Reply by yguclu** [reply]
That looks good, thanks.

I do not know if the name of the variable `starting_assertions` carries a meaning. If it does, please be careful that an assertion on an argument could be placed anywhere in the function, possibly inside an if condition or an if block (this happens quite often in Psydac, for example). Hence a bit of sophistication might be needed in the wrapper.

**Reply by EmilyBourne** [reply]
With `starting_assertions` I had in mind that this would only apply to the assertions placed at the top of the function. Otherwise we could be in trouble if the user does something like:
```python
def f(a : int):
    a *= 2
    assert a<40
```

It may be common to see
```python
def f(a : int = None):
    if a is not None:
        assert a < 4
```
But I don't think it is too much to ask people to write:
```python
assert a is None or a< 4
```
If they want wrapper error handling.
We don't want to end up with the entire function in the wrapper

**Reply by EmilyBourne** [reply]
```python
starting_assertions = []
i = 0
while isinstance(body[i], Assert):
     starting_assertions.append(body[i])
     i+=1
```
